pkg/v21/resource/appmigration/create.go

```diff
@@ -42,6 +47,100 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, chartSpec := range chartSpecsToMigrate {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if chartconfig CR %#q has been migrated", chartSpec.ChartName))
+
+		chartCR, err := getChartConfigByName(chartConfigs.Items, chartSpec.ChartName)
+		if IsNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("chartconfig CR %#q has been migrated, continuing", chartSpec.ChartName))
+			continue
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Cordon chartconfig CR so no changes are applied.
+		_, ok := chartCR.Annotations[annotation.CordonReason]
+		if !ok {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("cordoning chartconfig CR %#q", chartSpec.ChartName))
+
+			err = patchChartConfig(tenantG8sClient, chartCR, addCordonAnnotations())
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("cordoned chartconfig CR %#q", chartSpec.ChartName))
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("chartconfig CR %#q is already cordoned", chartSpec.ChartName))
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out app CR %#q is deployed", chartSpec.AppName))
+
+		appCR, err := r.g8sClient.ApplicationV1alpha1().Apps(clusterConfig.ID).Get(chartSpec.AppName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return microerror.Maskf(notFoundError, "app CR %#q", chartSpec.AppName)
+		}
+
+		if appCR.Status.Release.Status == "DEPLOYED" {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("app CR %#q has status %#q", chartSpec.AppName, appCR.Status.Release.Status))
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("adding annotation for deleting chartconfig CR %#q", chartSpec.ChartName))
+
+			// Add deletion annotation which will trigger chart-operator to
+			// delete the chartconfig CR but not the Helm release.
+			err = patchChartConfig(tenantG8sClient, chartCR, addDeleteAnnotation())
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("added annotation to chartconfig CR %#q", chartSpec.ChartName))
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting chartconfig CR %#q", chartSpec.ChartName))
+
+			// Lastly delete the chartconfig CR.
+			err = tenantG8sClient.CoreV1alpha1().ChartConfigs("giantswarm").Delete(chartCR.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted chartconfig CR %#q", chartSpec.ChartName))
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("app CR %#q has status %#q, continuing", chartSpec.AppName, appCR.Status.Release.Status))
+			continue
```

Comment:
@tomahawk28 I removed the retry logic we discussed in https://github.com/giantswarm/cluster-operator/pull/670.

I couldn't get it to work reliably without introducing a long max wait. There is no rush on the migration we can wait for the next resync period.